src/libostree/ostree-sign.c

```diff
@@ -0,0 +1,347 @@
+}
+
+OstreeSign * ostree_sign_get_by_name (const gchar *name, GError **error)
+{
+  g_debug ("%s enter", __FUNCTION__);
+
+  GType types [] = {
+#if defined(HAVE_LIBSODIUM)
+          OSTREE_TYPE_SIGN_ED25519,
+#endif
+          OSTREE_TYPE_SIGN_DUMMY
+  };
+  OstreeSign *ret = NULL;
+
+  for (gint i=0; i < G_N_ELEMENTS(types); i++)
+  {
+    g_autoptr (OstreeSign) sign = g_object_new (types[i], NULL);
+    g_autofree gchar *sign_name = OSTREE_SIGN_GET_IFACE (sign)->get_name(sign);
+
+    g_debug ("Found '%s' signing module", sign_name);
+
+    if (g_strcmp0 (name, sign_name) == 0)
+    {
+      ret = g_steal_pointer (&sign);
+      break;
+    }
+  }
+
+  if (ret == NULL)
+      g_set_error_literal (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+                           "Requested signature type is not implemented");
+
+  return ret;
+}
+
+
+/**
+ * ostree_sign_commit:
+ * @self: Self
+ * @commit_checksum: SHA256 of given commit to sign
+ * @cancellable: A #GCancellable
+ * @error: a #GError
+ *
+ * Add a GPG signature to a commit.
+ */
+gboolean
+ostree_sign_commit (OstreeSign     *self,
+                    OstreeRepo     *repo,
+                    const gchar    *commit_checksum,
+                    GCancellable   *cancellable,
+                    GError         **error)
+{
+  g_debug ("%s enter", __FUNCTION__);
+
+  g_autoptr(GBytes) commit_data = NULL;
+  g_autoptr(GBytes) signature = NULL;
+  g_autoptr(GVariant) commit_variant = NULL;
+  g_autoptr(GVariant) old_metadata = NULL;
+  g_autoptr(GVariant) new_metadata = NULL;
+
+  if (!ostree_repo_load_variant (repo, OSTREE_OBJECT_TYPE_COMMIT,
+                                 commit_checksum, &commit_variant, error))
+    return glnx_prefix_error (error, "Failed to read commit");
+
+  if (!ostree_repo_read_commit_detached_metadata (repo,
+                                                  commit_checksum,
+                                                  &old_metadata,
+                                                  cancellable,
+                                                  error))
+    return glnx_prefix_error (error, "Failed to read detached metadata");
+
+  // TODO: d4s: check if already signed?
+
+  commit_data = g_variant_get_data_as_bytes (commit_variant);
+
+  if (!ostree_sign_data (self, commit_data, &signature,
+                         cancellable, error))
+    return glnx_prefix_error (error, "Not able to sign the cobject");
+
+  new_metadata =
+    ostree_sign_detached_metadata_append (self, old_metadata, signature);
+
+  if (!ostree_repo_write_commit_detached_metadata (repo,
+                                                   commit_checksum,
+                                                   new_metadata,
+                                                   cancellable,
+                                                   error))
+    return FALSE;
+
+  return TRUE;
+}
+
+GStrv ostree_sign_list_names(void)
+{
+  g_debug ("%s enter", __FUNCTION__);
+
+  GType types [] = {
```

Comment:
This array should be global. No need to define it in every func needing the names. It could also contain the name in the form a string to avoid allocating an object just to read the name.